File: sdk/python/src/protomcp/tool.py
import inspect
import json
import types
import typing
from dataclasses import dataclass
from typing import Any, Callable, get_type_hints, get_origin, get_args

_registry: list["ToolDef"] = []

@dataclass
class ToolDef:
    name: str
    description: str
    input_schema_json: str
    handler: Callable
    output_schema_json: str = ""
    title: str = ""
    destructive_hint: bool = False
    idempotent_hint: bool = False
    read_only_hint: bool = False
    open_world_hint: bool = False
    task_support: bool = False
    hidden: bool = False

def tool(
    description: str,
    output_type=None,
    title: str = "",
    destructive: bool = False,
    idempotent: bool = False,
    read_only: bool = False,
    open_world: bool = False,
    task_support: bool = False,
    hidden: bool = False,
):
    def decorator(func: Callable) -> Callable:
        schema = _generate_schema(func)
        output_schema = _generate_dataclass_schema(output_type) if output_type is not None else {}
        _registry.append(ToolDef(
            name=func.__name__,
            description=description,
            input_schema_json=json.dumps(schema),
            handler=func,
            output_schema_json=json.dumps(output_schema) if output_schema else "",
            title=title,
            destructive_hint=destructive,
            idempotent_hint=idempotent,
            read_only_hint=read_only,
            open_world_hint=open_world,
            task_support=task_support,
            hidden=hidden,
        ))
        return func
    return decorator

def get_registered_tools() -> list[ToolDef]:
    from protomcp.group import groups_to_tool_defs
    from protomcp.workflow import workflows_to_tool_defs
    return list(_registry) + groups_to_tool_defs() + workflows_to_tool_defs()

def get_hidden_tool_names() -> list[str]:
    return [t.name for t in _registry if t.hidden]

def clear_registry():
    _registry.clear()

_PRIMITIVE_TYPE_MAP = {
    str: "string",
    int: "integer",
    float: "number",
    bool: "boolean",
    list: "array",
    dict: "object",
}

def _type_to_schema(hint) -> dict:
    origin = get_origin(hint)
    args = get_args(hint)
    if origin is typing.Union or origin is types.UnionType:
        non_none = [a for a in args if a is not type(None)]
        schemas = [_type_to_schema(a) for a in non_none]
        if type(None) in args:
            schemas.append({"type": "null"})
        if len(schemas) == 1:
            return schemas[0]
        return {"anyOf": schemas}
    if origin is list:
        if args:
            return {"type": "array", "items": _type_to_schema(args[0])}
        return {"type": "array"}
    if origin is dict:
        schema: dict[str, Any] = {"type": "object"}
        if len(args) == 2:
            schema["additionalProperties"] = _type_to_schema(args[1])
        return schema
    if origin is typing.Literal:
        value_types = set(type(v).__name__ for v in args)
        if len(value_types) == 1:
            t = type(args[0])
            json_type = _PRIMITIVE_TYPE_MAP.get(t)
            if json_type:
                return {"type": json_type, "enum": list(args)}
        return {"enum": list(args)}
    json_type = _PRIMITIVE_TYPE_MAP.get(hint)
    if json_type:
        return {"type": json_type}
    return {"type": "string"}

def _generate_schema(func: Callable) -> dict:
    hints = get_type_hints(func)
    sig = inspect.signature(func)
    properties = {}
    required = []
    for name, param in sig.parameters.items():
        if name in ("self", "cls", "ctx"):
            continue
        hint = hints.get(name, Any)
        if hasattr(hint, "__name__") and hint.__name__ == "ToolContext":
            continue
        prop = _type_to_schema(hint)
        if param.default is not inspect.Parameter.empty:
            prop["default"] = param.default
        elif not _is_optional_type(hint):
            required.append(name)
        properties[name] = prop
    schema: dict[str, Any] = {"type": "object", "properties": properties}
    if required:
        schema["required"] = required
    return schema

def _is_optional_type(hint) -> bool:
    origin = get_origin(hint)
    if origin is typing.Union or origin is types.UnionType:
        return type(None) in get_args(hint)
    return False

def _generate_dataclass_schema(dataclass_type) -> dict:
    import dataclasses
    if not dataclasses.is_dataclass(dataclass_type):
        return {}
    hints = get_type_hints(dataclass_type)
    fields = dataclasses.fields(dataclass_type)
    properties = {}
    required = []
    for field in fields:
        hint = hints.get(field.name, Any)
        prop = _type_to_schema(hint)
        if field.default is not dataclasses.MISSING or field.default_factory is not dataclasses.MISSING:  # type: ignore[misc]
            pass  # optional field
        elif not _is_optional_type(hint):
            required.append(field.name)
        properties[field.name] = prop
    schema: dict[str, Any] = {"type": "object", "properties": properties}
    if required:
        schema["required"] = required
    return schema

File: sdk/go/protomcp/tool.go
package protomcp

import (
	"encoding/json"
)

type ToolDef struct {
	Name         string
	Desc         string
	InputSchema  map[string]interface{}
	OutputSchema map[string]interface{}
	HandlerFn    func(ToolContext, map[string]interface{}) ToolResult
	Title        string
	Destructive  bool
	Idempotent   bool
	ReadOnly     bool
	OpenWorld    bool
	TaskSupport  bool
	Hidden       bool
}

type ToolOption func(*ToolDef)

var registry []ToolDef

func Tool(name string, opts ...ToolOption) {
	td := ToolDef{
		Name:        name,
		InputSchema: map[string]interface{}{"type": "object", "properties": map[string]interface{}{}},
	}
	for _, opt := range opts {
		opt(&td)
	}
	registry = append(registry, td)
}

func Description(desc string) ToolOption {
	return func(td *ToolDef) { td.Desc = desc }
}

type ArgDef struct {
	Name       string
	Type       string
	ItemType   string   // for array
	UnionTypes []string // for union
	EnumValues []string // for literal
}

func IntArg(name string) ArgDef  { return ArgDef{Name: name, Type: "integer"} }
func StrArg(name string) ArgDef  { return ArgDef{Name: name, Type: "string"} }
func NumArg(name string) ArgDef  { return ArgDef{Name: name, Type: "number"} }
func BoolArg(name string) ArgDef { return ArgDef{Name: name, Type: "boolean"} }

func ArrayArg(name string, itemType string) ArgDef {
	return ArgDef{Name: name, Type: "array", ItemType: itemType}
}

func ObjectArg(name string) ArgDef {
	return ArgDef{Name: name, Type: "object"}
}

func UnionArg(name string, types ...string) ArgDef {
	return ArgDef{Name: name, Type: "union", UnionTypes: types}
}

func LiteralArg(name string, values ...string) ArgDef {
	return ArgDef{Name: name, Type: "literal", EnumValues: values}
}

func argToSchema(a ArgDef) map[string]interface{} {
	switch {
	case a.Type == "array":
		schema := map[string]interface{}{"type": "array"}
		if a.ItemType != "" {
			schema["items"] = map[string]interface{}{"type": a.ItemType}
		}
		return schema
	case a.Type == "union" && len(a.UnionTypes) > 0:
		anyOf := make([]interface{}, len(a.UnionTypes))
		for i, t := range a.UnionTypes {
			anyOf[i] = map[string]interface{}{"type": t}
		}
		return map[string]interface{}{"anyOf": anyOf}
	case a.Type == "literal" && len(a.EnumValues) > 0:
		vals := make([]interface{}, len(a.EnumValues))
		for i, v := range a.EnumValues {
			vals[i] = v
		}
		return map[string]interface{}{"type": "string", "enum": vals}
	default:
		return map[string]interface{}{"type": a.Type}
	}
}

func Args(args ...ArgDef) ToolOption {
	return func(td *ToolDef) {
		props := map[string]interface{}{}
		required := []string{}
		for _, a := range args {
			props[a.Name] = argToSchema(a)
			required = append(required, a.Name)
		}
		td.InputSchema = map[string]interface{}{
			"type":       "object",
			"properties": props,
			"required":   required,
		}
	}
}

func Handler(fn func(ToolContext, map[string]interface{}) ToolResult) ToolOption {
	return func(td *ToolDef) { td.HandlerFn = fn }
}

func Title(v string) ToolOption                { return func(td *ToolDef) { td.Title = v } }
func DestructiveHint(v bool) ToolOption { return func(td *ToolDef) { td.Destructive = v } }
func IdempotentHint(v bool) ToolOption  { return func(td *ToolDef) { td.Idempotent = v } }
func ReadOnlyHint(v bool) ToolOption    { return func(td *ToolDef) { td.ReadOnly = v } }
func OpenWorldHint(v bool) ToolOption   { return func(td *ToolDef) { td.OpenWorld = v } }
func TaskSupportHint(v bool) ToolOption { return func(td *ToolDef) { td.TaskSupport = v } }
func HiddenHint(h bool) ToolOption      { return func(d *ToolDef) { d.Hidden = h } }

func (td ToolDef) InputSchemaJSON() string {
	b, _ := json.Marshal(td.InputSchema)
	return string(b)
}

func GetRegisteredTools() []ToolDef {
	all := append([]ToolDef{}, registry...)
	all = append(all, GroupsToToolDefs()...)
	all = append(all, WorkflowsToToolDefs()...)
	return all
}
func ClearRegistry()                { registry = nil }
